### PR TITLE
[AutoDiff][TF] Derivatives for reshaping methods

### PR DESCRIPTION
**Pull request by dave-fernandes**
Add derivative for `squeezingShape`. Fix derivative chaining for `reshaped` and `expandingShape`. Make convenience methods `flattened` and `rankLifted` differentiable. Fixes TF-367.

### CODE REVIEW

File: stdlib/public/TensorFlow/Tensor.swift [hunk 1]
```diff
@@ -710,6 +712,10 @@ public extension Tensor {
   // FIXME: The gradient for variadic `squeezed` is difficult to express because
```
Comment (dan-zheng) [inline stdlib/public/TensorFlow/Tensor.swift]: Please remove this deprecated comment. :)

File: stdlib/public/TensorFlow/Gradients.swift [hunk 2]
```diff
@@ -540,17 +540,28 @@ extension Tensor where Scalar : TensorFlowFloatingPoint {
...
-      return v.reshaped(toShape: self.shapeTensor)
+      return v.scalarCount == 1 ?
+        v.broadcast(toShape: self.shapeTensor) :
+        v.reshaped(toShape: self.shapeTensor)
     })
   }
 
+  @inlinable
+  func _vjpSqueezingShape(at axes: Int32...) -> (Tensor, (Tensor) -> Tensor) {
+    // Cannot call squeezingShape(at:) since it has variadic arguments
```
Comment (dan-zheng) [inline stdlib/public/TensorFlow/Gradients.swift]: Could you please create a non-variadic version `func squeezingShape(at axes: [Int32])` and define `_vjpSqueezingShape(at axes: [Int32])` for it?

File: test/TensorFlowRuntime/tensor_autodiff_runtime.swift [hunk 3]
```diff
@@ -117,6 +117,27 @@ TensorADTests.testAllBackends("mean") {
   expectEqual(expected, meanGradAlongAxes(input))
 }
 
+TensorADTests.testAllBackends("expandingShape") {
+  let f1 = { (a: Tensor<Float>) in a.expandingShape(at: 0).squared() }
+  let f2 = { (a: Tensor<Float>) in a.squared().expandingShape(at: 0) }
+  expectEqual([6, 10], gradient(at: [3, 5], in: f1))
```
Comment (dan-zheng) [inline test/TensorFlowRuntime/tensor_autodiff_runtime.swift]: `gradient` should be used only for functions that return scalar tensors (i.e. tensors with rank 0).
However, `f1` and `f2` do not return scalar tensors.

Please modify tests to call `pullback` instead, and pass in a direction vector of ones with the same shape as the output:
```suggestion
  expectEqual([6, 10], pullback(at: [3, 5], in: f1)([[1, 1]]))
  expectEqual([6, 10], pullback(at: [3, 5], in: f2)([[1, 1]]))
```

File: stdlib/public/TensorFlow/Gradients.swift [hunk 4]
```diff
@@ -540,17 +540,28 @@ extension Tensor where Scalar : TensorFlowFloatingPoint {
   ) -> (Tensor, (Tensor) -> Tensor) {
     let value = reshaped(toShape: newShape)
     return (value, { v in
-      return v.reshaped(toShape: self.shapeTensor)
+      return v.scalarCount == 1 ?
```
Comment (dan-zheng) [inline stdlib/public/TensorFlow/Gradients.swift]: Please revert this change.

Shape mismatches were caused by calling `gradient` on functions that don't return scalar tensors. Instead of adding a special derivative definition that performs broadcasting, the correct change is to call `pullback` with a direction vector of the correct shape (same shape as function result).

Let me know if this makes sense!

File: stdlib/public/TensorFlow/Gradients.swift [hunk 5]
```diff
@@ -540,17 +540,28 @@ extension Tensor where Scalar : TensorFlowFloatingPoint {
...
+  }
+
   @inlinable
   func _vjpExpandingShape(
     at shapeIndex: Int32
   ) -> (Tensor, (Tensor) -> Tensor) {
     let value = expandingShape(at: shapeIndex)
     return (value, { v in
-      return v.squeezingShape(at: shapeIndex)
+      return v.scalarCount == 1 ?
```
Comment (dan-zheng) [inline stdlib/public/TensorFlow/Gradients.swift]: Similarly, please revert this change.

File: stdlib/public/TensorFlow/Gradients.swift (hunk 4 above)
Comment (dave-fernandes) [reply stdlib/public/TensorFlow/Gradients.swift]: Ah, okay. I was misunderstanding how gradient could be used.

**Comment by dave-fernandes** [comment]
Thanks for reviewing @dan-zheng. I have address all the feedback.

File: stdlib/public/TensorFlow/Gradients.swift [hunk 6]
```diff
@@ -544,6 +544,14 @@ extension Tensor where Scalar : TensorFlowFloatingPoint {
     })
   }
 
+  @inlinable
+  func _vjpSqueezingShape(at axes: [Int32]) -> (Tensor, (Tensor) -> Tensor) {
+    let value = squeezingShape(at: axes)
+    return (value, { v in
+      return v.reshaped(toShape: self.shapeTensor)
```
Comment (rxwei) [inline stdlib/public/TensorFlow/Gradients.swift]: * For better memory efficiency, it would be nice to not capture `self` in the pullback closure when it's not needed. 
* Omit `return` when a closure body has only one expression.

You can define the pullback this way
```swift
return (value, { [shape = shapeTensor] v in
  v.reshaped(toShape: shape)
})
```

**Review by rxwei — APPROVED** [review]
Thanks Dave!

File: stdlib/public/TensorFlow/Gradients.swift (hunk 6 above)
Comment (dave-fernandes) [reply stdlib/public/TensorFlow/Gradients.swift]: I'll make that change and make _vjpReshaped and _vjpExpandingShape consistent as well. There seem to be a couple different idioms used in this file.

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

File: stdlib/public/TensorFlow/Gradients.swift (hunk 6 above)
Comment (rxwei) [reply stdlib/public/TensorFlow/Gradients.swift]: Thanks!

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow
